### ISSUE TITLE
GUI Redesign request (Discussion)

### ISSUE THREAD

**Issue by aha999** [fr3ts0n/AndrOBD#104]
I think that the AndrOBD is awesome by being open source and supporting all the elm327, but the GUI doesn't look too good or modern at the moment and i hope that there are contributors that know this stuff.

This issue can serve as a discussion for GUI redesign among contributors / devs (if you don't close it).

What is your standpoint, main dev @fr3ts0n ?

**Comment by fr3ts0n** [comment]
Yes, it is quite obvious that GUI design of AndrOBD is not my strong spot.

However, I have always been kind of proud to deliver a functional OBD app in <500KB size.
I tried to keep functionality before "fancyness" and eliminated all redundant resources.
I still want to keep AndrOBD to be a lean app, but totally agree that the UI has quite some potential of improvements.

I would appreciate to get some helping hands on improvements of the GUI.

**Comment by aha999** [comment]
Examples of other obd apps GUI that i like, you can search for similar apps on google play by searching "OBD", i reviewed some and selected the better looking ones:
[First app](https://play.google.com/store/apps/details?id=com.ovz.carscanner), [second app](https://play.google.com/store/apps/details?id=OCTech.Mobile.Applications.OBDLink), [third app](https://play.google.com/store/apps/details?id=com.palmerperformance.DashCommand), [fourth app](https://play.google.com/store/apps/details?id=com.lemurmonitors.bluedriver), [fifth app](https://play.google.com/store/apps/details?id=com.obdautodoctor), [sixth app](https://play.google.com/store/apps/details?id=com.elm.obd.harry.scan), [seventh app](https://play.google.com/store/apps/details?id=it.smartapps4me.smartcontrol), [Eighth app](https://play.google.com/store/apps/details?id=com.GynetaGroup.CarDiagnosticsLite), [Ninth app](https://play.google.com/store/apps/details?id=com.GynetaGroup.CarDiagnostics)

**Comment by aha999** [comment]
This might help:

- https://material.io/
- https://developer.android.com/design
- https://developer.android.com/guide/topics/ui

**Comment by brunozaranza** [comment]
I'm android developer and I'd happy to help about this, cuz I'dont able to understand how use this app. BTW, how to do anything after connecting the OBD? The screen is blank.

**Comment by aha999** [comment]
@brunozaranza check out [AndrOBD chat ](https://t.me/joinchat/G60ltQv5CCEQ94BZ5yWQbg) on telegram.
 That would be awesome if you could help! Please look at the posted GUI related issues. After the connection The screen is blank but options are currently in the top right corner.

**Comment by aha999** [comment]
@brunozaranza yeah imo the app is visually not polished and wanted measurements are hard to find since it's so many of them in one list. After Connecting the OBD (or opening demo mode), you can access info by pressing two options in the top bar in the app, which will open drop down menus.

**Comment by aha999** [comment]
@fr3ts0n i think it would increase the chances of someone helping if you posted this issue to https://github.com/up-for-grabs/up-for-grabs.net so it would be visible on their website

**Comment by fr3ts0n** [comment]
Dear potential contributors,

Thank you very much for your interest in re-freshing the AndrOBD GUI design.

There have been several questions concerning the used programming language of **Java**.
The AndrOBD project **also supports implentations in Kotlin** as well, so this is also a valid option to be agreed on.
[See FAQ](https://github.com/fr3ts0n/AndrOBD/wiki/Frequently-asked-questions#what-programming-languages-can-be-used-for-contributions)

However, there are a few steps we should agree on, before someone takes off and silently spends much effort to create a new GUI.

- It would make a lot of sense if you could provide an early, rough draft design of your GUI ideas before spending much effort in implementations.
  - This could just even be some paper drafts on screen layouts.
- We could then do some review discussions of your ideas, and find agreements on open questions.
  - I would very well appreciate to have @aha999 join these reviews, since I think that he has a rather good knowledge & feel for usability. 
  - I also would like to encourage additional, interested attendants to these review sessions.
    -  but I understand if you prefer to review your ideas in a smaller round only.
- With the current interface I frequently did use resources from the Android system, rather than adding some custom ones, to keep the implementation as lean as possible.
  - This might be just a personal preference, but I hate apps which use up many megabytes of memory just to look fancy without any functional purpose.
  - It makes me kind of proud to fit all of AndrOBD's functionality into a 1.4MB app.
- Let's try to join our forces to create one good compromise of functional, well usable, fancy looking but reasonable lean GUI.

Challenge accepted?

Then let's go for it ...

I'm looking forward to hear/read your design ideas 👍

**Comment by GiovanniOliveira75** [comment]
Hi everyone. @fr3ts0n I am a software engineer and UX/UI designer and I would like to contribute with designs to this project. I was looking for an OBD solution a couple of months ago and this project is the best so far. Can I prepare something to show you?

**Comment by fr3ts0n** [comment]
Hi @GiovanniOliveira75 ,
You are very welcome to prepare some proposal.
I am  looking forward to see your ideas.

**Comment by Gamesnic** [comment]
Hey there,

Can you only make use of actual code to put into the app or also prototype/concept of how it could look?

**Comment by pratikraman-dev** [comment]
I'm android developer and I'd happy to help @fr3ts0n  !

**Comment by fr3ts0n** [comment]
Hi @pratikraman-dev ,
Thanks for your offer to help. 

We are very keen on your initial drafts...

**Comment by peterleequigley** [comment]
Hey guys,

Looks like you've had a few takers in the past few months. Are you still looking for interface redesign? Looks like you provide a real, singular useful service right now, and I definitely have bandwidth to contribute to a GUI redesign.

I have a bunch of questions though. Is there a way to connect for kind of a groundwork conversation? Feel free to email me at peterleequigley@gmail.com, or message me on Discord at Kirin#6171. Thanks!

**Comment by aha999** [comment]
@GiovanniOliveira75 I just created a Discord server with the permission of @fr3ts0n. The link is here: https://discord.gg/wdZ5eYYFyD . As the server is fresh, it will get tidied up :)

**Comment by peterleequigley** [comment]
@fr3ts0n I just saw the Discord invite, looks like a good start! I'll copy this post over to there as well.

Couple things I'd like to do to get started:
- **Get familiar with AndrOBD**. It'd be helpful to see a demo of AndrOBD, since I don't have as much of the context. (I tried downloading it myself via F-Droid, but I don't think it's supported on newer Androids?) This will also give a chance to get basic questions out of the way and make sure I don't go off and design something that doesn't align properly with the app.

Following that, I can start researching and developing some early designs. I'll want to try and talk with some users of the app at that point to find out what their main priorities are, too.

If that all sounds good to you, feel free to email me (peterleequigley@gmail.com), or just respond here or in Discord. Thanks!

**Comment by fr3ts0n** [comment]
Current Versions of Android should definitely be supported.

Demo Mode will be started If you do NOT allow the App to connect to BT.
See: https://github.com/fr3ts0n/AndrOBD/wiki/Frequently-asked-questions#activation-of-androbd-demo-mode

**Comment by peterleequigley** [comment]
Yup, it definitely works! (My device had been blocking F-Droid downloads)

I could probably put together an improved UI based on the app and the content I see on the website, but it might suffer because I don't know much about OBD scanning or car diagnostics. I could get much farther if I were able to talk to some actual users, or people that are super familiar with the domain that can clue me in.

We would also need a **front-end developer** to actually apply any designs we end up with. It could also be possible to just put together some designs now, and convert the task to an unassigned Github task for a front-end developer. Let me know your thoughts.

**Comment by aha999** [comment]
@peterleequigley you can take a look at some of the[ alternative apps,](https://github.com/fr3ts0n/AndrOBD/issues/104#issuecomment-576083789) to get the idea of the content that is important.

Discord group is now available at [AndrOBD - Discord](https://discord.gg/wdZ5eYYFyD).

**Comment by aha999** [comment]
@peterleequigley you can take a look at some of the[ alternative apps,](https://github.com/fr3ts0n/AndrOBD/issues/104#issuecomment-576083789) to get the idea of the content that is important.

Discord group is now available at [AndrOBD - Discord](https://discord.gg/wdZ5eYYFyD).

**Comment by kaustavr19** [comment]
I am a Product Designer and would love to redesign the application for you. If you are okay with that, kindly let me know how to get started. Thanks.

**Comment by fr3ts0n** [comment]
Hi @kaustavr19 , 

Thank you for your offer. 
Yes, you are very welcome to provide ideas for this redesign.
But just to make it clear up front, unfortunately AndrOBD can not pay you anything for this, since it is a NON profit OSS project.

As mentioned in my [earlier post](https://github.com/fr3ts0n/AndrOBD/issues/104#issuecomment-1106363617) it would be helpful to discuss early drafts before implementation. 

If that is OK with you, I would suggest you can provide your draft changes with a pull request, and we can start discussions there.

Looking forward to see your ideas ...

**Comment by MOtterbine** [comment]
...interesting discussion

**Comment by aha999** [comment]
How can we get a good UX and UI app design, can we post a "bounty"?
I can contribute 5€ to the design and 5€ to the development of a new UI.

**Comment by treysblackshoes** [comment]
@fr3ts0n 
> However, I have always been kind of proud to deliver a functional OBD app in <500KB size.
I tried to keep functionality before "fancyness" and eliminated all redundant resources.

You should be really proud of this. There's a ton of apps out there (not just OBD apps) that could be more functional and reliable, but instead resources are wasted on fancy graphics. My advice on the redesign is to continue using Material 2 for now. It's consistent, unlike Google's plans for Material 3, but especially its new "expressive" variant, which looks so confusing I'm convinced it's a publicity stunt for investors.

**Comment by ThatOneCalculator** [comment]
IMO M3/M3E looks great, is easier to build in, and apps made with it tend to feel really good compared to M2 apps

**Comment by unfa** [comment]
Hey I am a user of the app and a graphic designer. My specialty is 3D and shaders, but I also did stuff like designing car dashboard clocks etc at work (game development).

I am not a Java or Kotlin developer and I can't help on the code side, but I could lend my aesthetic sense and make assets in SVG for example I'd there's a need for some.

If there's a need for sound, I could also design bespoke sound effects for this app as I have done quite a bit of that kind of work in the past, and it's something I enjoy doing.

Thank you for this wonderful libre tool! I can't stand the stuff that Google Play is stuffed with, and I would love to give back to this project if I can.

**Comment by chrisdebian** [comment]
Hi @fr3ts0n,

I've been contributing some small fixes to AndrOBD recently and took the opportunity to do a code-level UI audit before suggesting anything concrete for this thread. Here's what I found, and a staged proposal if you're interested.

## What the audit found

The app currently has no Material Design library dependency at all — the theme inherits from `@android:style/Theme.DeviceDefault`, which is the Holo-era Android 4.x system theme. This means the "dated" look reported here isn't a Material 2 → Material 3 gap: it's starting from zero. The good news is that the first step (adding the library and migrating the theme) is small and low-risk.

A few other findings worth flagging:

- **AChartEngine** (used for the chart view) appears to have been effectively abandoned since around 2013. This isn't urgent, but it's a long-term maintenance risk.
- **The biggest discoverability problem** is navigation, not visuals. OBD Services — the primary feature — is hidden in the overflow menu and starts greyed-out until connected. Several comments in this thread mention the "blank screen" confusion; this is why. A bottom navigation bar would solve it without any visual redesign.
- All lists use `ListView`/`GridView` rather than `RecyclerView`.
- Several layouts have hardcoded hex colours that break in certain system themes.

## Proposed staged approach

I'd suggest tackling this in phases so each change is reviewable on its own and nothing gets blocked waiting for a complete redesign:

**Phase 1 — Theme baseline** *(small PR, low risk)*
Add `androidx.appcompat` and `com.google.android.material` as dependencies; migrate `styles.xml` from `Theme.DeviceDefault` to `Theme.Material3.DayNight`. No layout changes. The app would look similar but gain proper typography scaling, dynamic colour support on Android 12+, and a foundation for everything that follows. APK size impact is roughly 200–300 KB depending on shrinking configuration.

**Phase 2 — Icons and controls** *(medium, screen by screen)*
Replace `@android:drawable/ic_menu_*` Holo icons with Material Symbols equivalents; migrate `ProgressBar` → `CircularProgressIndicator`, plain `Button` → `MaterialButton`, `SeekBar` → `Slider`. One PR per screen.

**Phase 3 — Navigation** *(larger, needs discussion)*
Move OBD service selection to a `BottomNavigationView` so it's always visible and clearly enabled/disabled based on connection state. This is the highest-impact change for usability but also the most significant structural change — worth a separate discussion before starting.

**Phase 4 — Lists and chart** *(largest)*
Migrate `ListView`/`GridView` to `RecyclerView`; replace AChartEngine with a maintained charting library.

---

I'm happy to start with Phase 1 if this direction sounds right to you. I wanted to check in before writing any code given the scope, and also because you've mentioned the importance of keeping the APK lean — worth agreeing on the ProGuard/R8 shrinking configuration for the Material dependency before we add it.

Thanks, Chris

**Comment by fr3ts0n** [comment]
Hi @chrisdebian ,
Thank you for your audit and proposal.

I like the phased approach suggestion. The APK size impact also sounds very reasonable to me.
How about backwards compatibility? What will be the targeted minimum version of Android?
Referring to the [usage query](https://t.me/c/1434285001/2986) on Telegram there is approx. 25% users of Android <= 12.

Thx

**Comment by chrisdebian** [comment]
Good question. Checked this directly against the Material Components library's own build config — it currently requires minSdk 23 (Android 6.0), so adopting it would mean bumping AndrOBD's floor from 17→23. That only drops Android 4.2–5.1 (API 17–22) — devices over a decade old now, almost certainly a negligible slice of the active user base.

Importantly, this is separate from your 25% figure: those are Android ≤ 12 users, and they're not affected by the minSdk bump at all — Material 3 styling works fully via the AppCompat backport all the way down to API 23. The one thing that's Android-12-specific is *dynamic* colour (deriving the theme from the user's wallpaper) — that's an OS feature requiring API 31, not a library requirement. Below that, the app would just use a fixed/static palette instead, which was already the planned fallback — so pre-Android-12 users still get the full modernised look, just without wallpaper-matched theming.

So: no impact on your 25% Android ≤ 12 group either way, and only a very old, likely tiny slice of users would be dropped by the minSdk bump itself. Happy to start with Phase 1 if that sounds like an acceptable trade-off.

**Comment by fr3ts0n** [comment]
For the Users on <= 5.1 (API 22) there most likely would have been an end of support/updates by SDK update requirements anyway. 🥹 
I think we can definitely start with Phase1 then.

I will create a release with the latest fixes before any of the updates, so all the 'old' version can still profit from it.

**Comment by chrisdebian** [comment]
Wanted to correct something before going further: I actually built this now (finally got a working local Android SDK setup) and the real APK size impact is bigger than I estimated — sorry for the bad number earlier.

Measured with a clean build, master vs. adding just the AppCompat + Material dependencies and swapping the theme parent (no other changes):

| | master | +Material/AppCompat |
|---|---|---|
| debug | 2.8 MB | 6.8 MB |
| release (unsigned) | 2.3 MB | 5.6 MB |

So it's roughly **+3.3 MB even in the release build**, not the 200-300 KB I originally said. The reason: this project doesn't currently have `minifyEnabled`/R8 shrinking configured anywhere (no `buildTypes` block at all), so nothing gets stripped from either dependency — my original estimate assumed shrinking would apply, which it doesn't yet.

Two ways forward, your call:
1. **Accept the larger unshrunk size** for now — Phase 1 still compiles clean and works, just heavier than promised.
2. **Set up R8 shrinking first** as its own small prerequisite PR, then re-measure — likely brings it much closer to the original estimate, since most of Material/AppCompat's code and resources would get stripped as unused. This is separate, real work though, and needs care given reflection-based code elsewhere in the app (USB serial, etc.) — R8 can silently break reflection if keep rules aren't right.

**Comment by chrisdebian** [comment]
Opened #335 for Phase 1 (theme baseline, minSdk 17→23) — includes the honest size numbers from my last comment here. Holding it open rather than waiting, since the release has shipped; happy to sit on merge until we've settled the size question.

**Comment by fr3ts0n** [comment]
What would be your preference?
I'd like to suggest to go with the bigger size right now and do the cleanup afterwards.
Let's keep an issue open to do th R8 shrinking after the visual / functional stuff.
Is that ok with you?

**Comment by chrisdebian** [comment]
Yes, that works well — go with the bigger size now, and we'll track R8 shrinking as its own follow-up once the visual/functional work has settled. I'll open a tracking issue for it now.

**Comment by chrisdebian** [comment]
Opened #339 to track the R8 shrinking follow-up.

**Comment by chrisdebian** [comment]
Now that Phase 1 (#335) has landed, happy to keep going whenever you're ready — no rush at all.

For Phase 2, I'd suggest starting with the smallest possible step rather than a big batch: the progress dialog (`ModernProgressDialog`) currently uses a plain `ProgressBar`. Swapping it for Material's `CircularProgressIndicator` is a one-screen, behaviour-identical change — easy to review quickly and a good way to confirm the Material dependency is working well in practice before touching anything bigger.

If that sounds reasonable, I'll open it as a PR. Otherwise, happy to start somewhere else on the list from the original audit, or hold off entirely if now isn't the right time — your call either way.

**Comment by fr3ts0n** [comment]
With current Integration status I don't see any title/menu bar, so no function is selectable.
Would it make sense to get a menu first?

**Comment by chrisdebian** [comment]
That's fine by me — ship the functional/visual progress now, track R8 shrinking as its own follow-up issue. No point blocking Phase 1 on an optimisation that's independent of the actual redesign work. Happy for you to open that tracking issue whenever suits.

**Comment by chrisdebian** [comment]
Good catch, and I think I've found why. PR #335 only touched `build.gradle` and `styles.xml` — it didn't touch any Activity classes. `Theme.Material3.*` is a Material Components/AppCompat theme, and its action bar support is wired through `AppCompatActivity`'s delegate. But every activity in the app (`MainActivity` via `PluginManager`, `BtDeviceListActivity`, `DashBoardActivity`, `UsbDeviceListActivity`, `ChartActivity`) currently extends plain `Activity`, not `AppCompatActivity` — so the theme has nothing to hook the action bar/menu into.

So yes, agreed: a menu first makes sense, but I'd frame it slightly differently — this isn't really "Phase 2 polish before a menu," it's that migrating the activities to `AppCompatActivity` (with a `Toolbar`) is a missing prerequisite that should have been part of Phase 1's theme baseline, not a separate step afterwards. Want me to put together a PR for that specifically — just the `AppCompatActivity` + `Toolbar` migration, no other visual changes — so the Material3 theme actually has an action bar to render into?

**Comment by fr3ts0n** [comment]
That would be just perfect. :+1:

**Comment by chrisdebian** [comment]
Opened #341 for the AppCompatActivity migration on `BtDeviceListActivity`, `DashBoardActivity`, `UsbDeviceListActivity`, and `ChartActivity`.

One thing I found while scoping this: `MainActivity` doesn't extend `Activity` directly — it extends `PluginManager` from `AndrOBD-libplugin`, which itself extends the deprecated `ListActivity`. So `MainActivity` can't get the same fix without also migrating `PluginManager` in that repo. Wanted to flag it before assuming scope on your other project — happy to look at it there too if that's welcome, or leave it as a `ListActivity`-based exception if you'd rather keep the plugin manager separate. Your call.

**Comment by fr3ts0n** [comment]
Yes, MainActivity derives from PluginManager. So it just makes sense to do that as well.
Sorry for increasing the scope

**Comment by fr3ts0n** [comment]
You do an incredible job.
Thank you so much for your help and deep Android knowledge.

**Comment by chrisdebian** [comment]
No apology needed at all — better to fix it properly than leave a `ListActivity` exception lying around. I'll take a look at `AndrOBD-libplugin` and come back with a scoped proposal for migrating `PluginManager` off `ListActivity`, same pattern as the four activities in #341. And thank you — really enjoying working on this with you.

**Comment by chrisdebian** [comment]
Opened the PluginManager migration this depends on: fr3ts0n/AndrOBD-libplugin#36. Scoping it turned up more `getListView()`/`getListAdapter()` usage in `MainActivity` than expected (~22 call sites), but re-adding those as plain compatibility methods on `PluginManager` itself meant `MainActivity` needs no changes at all — confirmed with a full app build against the branch.

**Comment by chrisdebian** [comment]
Both prerequisite PRs are merged now — #341 (AndrOBD) and fr3ts0n/AndrOBD-libplugin#36 (PluginManager) — so every activity including MainActivity now has a working action bar. That closes out the whole prerequisite chain.

Ready to move on to the first real Phase 2 step whenever you're happy: swapping the `ProgressBar` in the progress dialog for a Material `CircularProgressIndicator`, as proposed earlier — smallest possible step to start with. Let me know if you'd like me to go ahead.

**Comment by fr3ts0n** [comment]
Fine with me. 
Go ahead if you like. 
Thx

**Comment by chrisdebian** [comment]
Opened the first Phase 2 step as its own small PR: #343 (ProgressBar → CircularProgressIndicator in the progress dialog).

**Comment by chrisdebian** [comment]
Thanks for merging #343.

Next queued item per the suggested order is Settings (`SettingsActivity`): migrating from `PreferenceActivity` (deprecated since API 28) to `PreferenceFragmentCompat` with Material theming. Low effort, and fixes a real deprecation rather than just a cosmetic change.

Happy to go ahead whenever you're ready, same as before.

**Comment by fr3ts0n** [comment]
This is fine with me.

**Comment by chrisdebian** [comment]
Opened #344 for the Settings screen migration (item 2 in the suggested order) — PreferenceActivity → PreferenceFragmentCompat.

**Comment by chrisdebian** [comment]
Thanks for merging #344.

Next queued item per the suggested order is icons throughout (item 3) — swapping the Holo `@android:drawable/ic_menu_*` icons for Material Symbols. Touches every screen, so I'd do it as its own dedicated pass rather than folding it into another screen's PR.

Happy to go ahead if that works for you.

**Comment by fr3ts0n** [comment]
Thx.
That should work.

However, with #344 I can See a list of setting groups only, but selecting each does not open any sub-settings. How do you currently perform functional tests with your changes?

**Comment by chrisdebian** [comment]
Opened #345 to fix this — root cause was AndroidX Preference not auto-navigating into nested PreferenceScreens the way the old PreferenceActivity/headers API did. Testing surfaced two more bugs in the same area (a crash entering any screen, and a layout glitch on returning from one), both fixed in the same PR. Verified on-device across all groups and back-navigation depths.

**Comment by fr3ts0n** [comment]
@chrisdebian 
Just to let you know,
I will be on a vacation trip the next 10 days, so I may not be very responsive during that time.
Since you currently do all the work, pls don't get mad at me if I don't react in time ... 
Thx

**Comment by chrisdebian** [comment]
Enjoy the trip! No need to worry about response time — I'll keep queuing up well-scoped work per what's already agreed (starting with the icon migration from the 15 July thread), and hold off on anything that needs your judgement call until you're back. Thanks for everything so far, this has been genuinely enjoyable to work on.

Thanks, Chris

**Comment by chrisdebian** [comment]
Icon migration (Phase 2 item 3) is up: #347 on the main app, plus a companion #37 on AndrOBD-libplugin (its own settings.xml needs its own icon resources, since it's built independently).

Found a real light-mode visibility bug in the 4 existing custom icons along the way (hardcoded translucent-white fill, invisible on the app's white background) — fixed those too, all icons now use a themeable colour. Verified on-device in both light and dark mode.

**Comment by fr3ts0n** [comment]
@chrisdebian 
Do you currently use any local changes for testing the app?
Somehow I do not get any ToolBar / menu with the current integraton, so I can not select any function ...

**Comment by fr3ts0n** [comment]
@chrisdebian 
I had to disable **MissingTranslation** errors in linting, since this causes errors with most incomong translation changes.

**Comment by chrisdebian** [comment]
Sorry for the slow reply here — this should already be resolved. The blank-toolbar/no-menu issue was a real bug (`PluginManager.listView` never got re-resolved after `AppCompatActivity`'s `setContentView`), fixed in AndrOBD-libplugin#39, merged 2026-08-01 — one day before this comment. This repo's `plugin` submodule pointer has since been bumped past that fix.

If you're still seeing a blank Toolbar on a fresh clone/pull, let me know and I'll dig further — otherwise this should be working now.

**Comment by chrisdebian** [comment]
Item 5 (BT/BLE/USB device pickers) is up: #360.

**Comment by fr3ts0n** [comment]
@chrisdebian 
Thx, I found the blank toolbar issue to be a repo config problem on my end. 
So it is not an issue any more 👍
